tests/kafka.rs

```diff
@@ -0,0 +1,80 @@
+use futures::{StreamExt, TryStreamExt};
+use rdkafka::{
+    config::ClientConfig,
+    consumer::{stream_consumer::StreamConsumer, Consumer},
+    producer::{FutureProducer, FutureRecord},
+    Message,
+};
+use std::time::Duration;
+
+use testcontainers::{clients, core::RunArgs, images::kafka};
+
+#[tokio::test]
+async fn test_produce_and_consume_messages() {
+    let docker = clients::Cli::default();
+    let kafka_node = docker.run_with_args(
+        kafka::Kafka::default(),
+        RunArgs::default().with_mapped_port((kafka::KAFKA_PORT, kafka::KAFKA_PORT)),
+    );
+
+    let bootstrap_servers = format!("localhost:{}", kafka_node.get_host_port(kafka::KAFKA_PORT));
+
+    let producer: FutureProducer = ClientConfig::new()
+        .set("bootstrap.servers", &bootstrap_servers)
+        .set("message.timeout.ms", "5000")
+        .create()
+        .expect("Failed to create Kafka FutureProducer");
+
+    let consumer: StreamConsumer = ClientConfig::new()
+        .set("group.id", "testcontainer-rs")
+        .set("bootstrap.servers", &bootstrap_servers)
+        .set("session.timeout.ms", "6000")
+        .set("enable.auto.commit", "false")
+        .set("auto.offset.reset", "earliest")
+        .create()
+        .expect("Failed to create Kafka StreamConsumer");
+
+    let topic = "test-topic";
+    consumer
+        .subscribe(&[&topic])
+        .expect("Failed to subscribe to a topic");
+
+    let number_of_messages_to_produce = 5_usize;
+    let expected: Vec<String> = (0..number_of_messages_to_produce)
+        .map(|i| format!("Message {}", i))
+        .collect();
+
+    for (i, message) in expected.iter().enumerate() {
+        producer
+            .send(
+                FutureRecord::to(&topic)
+                    .payload(message)
+                    .key(&format!("Key {}", i)),
+                Duration::from_secs(0),
+            )
+            .await
+            .unwrap();
+    }
+
+    let actual = consumer
+        .stream()
+        .take(number_of_messages_to_produce)
+        .map_ok(|borrowed_message| {
+            // It is safe to unwrap as the producer has included payloads above.
+            borrowed_message
+                .payload_view::<str>()
+                .unwrap()
+                .unwrap()
+                .to_owned()
+        })
+        .collect::<Vec<_>>()
+        .await;
+
+    let number_of_messages_consumed = actual
+        .iter()
+        .zip(expected.iter())
+        .filter(|&(consumed, produced)| consumed.as_ref().unwrap() == produced)
+        .count();
+
+    assert_eq!(number_of_messages_to_produce, number_of_messages_consumed);
```

Comment:
This assertion is not particularly robust.

Let's say something actually goes wrong and we don't receive as many messages as we publish. A stream with `take(number_of_messages_to_produce).collect()` will wait forever to receive `number_of_messages_to_produce` messages. Hence, we are never actually triggering the final assertion.

However, I guess we can't collect the whole stream because it will never actually end because there could always be more messages.

I think the most robust approach would be to do the assertion something like this (needs proper type conversions and error handling):

```rust
let mut produced_messages = consumer.stream();

for msg in expected {
	let produced = tokio::time::timeout(produced_messages.next(), Duration::from_secs(2)).unwrap();

	assert_eq!(msg, produced)
}
```

Reply:
It's a great suggestion. I think it's important to check timeout issues as part of verification. 81be19a addressed it and I intentionally use `unwrap` throughout. I learned that testing is one of the places where it makes sense to use `unwrap`. Let me know if we want to use alternatives like `expect`.